server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -203,15 +203,16 @@ private static boolean noTemplateRuleWithOrganizationsEnabled(RegisterRulesConte
     private final Set<RuleDefinitionDto> unchanged = new HashSet<>();
     private final Set<RuleDefinitionDto> removed = new HashSet<>();
 
-    private RegisterRulesContext(Map<RuleKey, RuleDefinitionDto> dbRules, Map<Integer, Set<SingleDeprecatedRuleKey>> dbDeprecatedKeysById) {
+    private RegisterRulesContext(Map<RuleKey, RuleDefinitionDto> dbRules, Set<SingleDeprecatedRuleKey> dbDeprecatedKeys) {
       this.dbRules = ImmutableMap.copyOf(dbRules);
       this.known = ImmutableSet.copyOf(dbRules.values());
-      this.dbDeprecatedKeysById = dbDeprecatedKeysById;
-      this.dbRulesByDbDeprecatedKey = buildDbRulesByDbDeprecatedKey(dbDeprecatedKeysById, dbRules);
+      this.dbDeprecatedKeys = dbDeprecatedKeys;
+      this.dbRulesByDbDeprecatedKey = buildDbRulesByDbDeprecatedKey(dbDeprecatedKeys, dbRules);
     }
 
-    private static Map<RuleKey, RuleDefinitionDto> buildDbRulesByDbDeprecatedKey(Map<Integer, Set<SingleDeprecatedRuleKey>> dbDeprecatedKeysById,
-      Map<RuleKey, RuleDefinitionDto> dbRules) {
+    private static Map<RuleKey, RuleDefinitionDto> buildDbRulesByDbDeprecatedKey(Set<SingleDeprecatedRuleKey> dbDeprecatedKeys, Map<RuleKey, RuleDefinitionDto> dbRules) {
+      Map<Integer, Set<SingleDeprecatedRuleKey>> dbDeprecatedKeysById = dbDeprecatedKeys.stream()
```

Comment:
this map is not required to build rulesByDbDeprecatedKey. 
Loop at line 220 can easily be changed to work directly with the `Set<SingleDeprecatedRuleKey>`